cluster-autoscaler/ca.patch

```diff
@@ -1,137 +1,66 @@
+@@ -221,6 +228,22 @@ type AllPodLister struct {
+ 	podLister v1lister.PodLister
+ }
+ 
++func updatePodRequestsFromNeonVMAnnotation(pod *apiv1.Pod) {
++	annotation, ok := pod.Annotations["vm.neon.tech/usage"]
++	if !ok {
++		return
 +	}
 +
-+	return &cache.ListWatch{
-+		ListFunc: func(options metav1.ListOptions) (runtime.Object, error) {
-+			obj, err := lw.List(options)
-+			if err != nil {
-+				return obj, err
-+			}
-+
-+			list := obj.(*apiv1.PodList)
-+			for i := range list.Items {
-+				updatePodRequestsFromNeonVMAnnotation(&list.Items[i])
-+			}
-+			return obj, nil
-+		},
-+		WatchFunc: func(options metav1.ListOptions) (watch.Interface, error) {
-+			iface, err := lw.Watch(options)
-+			if err != nil {
-+				return iface, err
-+			}
-+
-+			// Wrap the channel to update the pods as they come through
-+			wrappedEvents := make(chan watch.Event)
-+			proxyIface := watch.NewProxyWatcher(wrappedEvents)
-+
-+			go func() {
-+				events := iface.ResultChan()
-+
-+				for {
-+					var ok bool
-+					var ev watch.Event
-+
-+					select {
-+					case <-proxyIface.StopChan():
-+						return
-+					case ev, ok = <-events:
-+						if !ok {
-+							close(wrappedEvents)
-+							return
-+						}
-+					}
-+
-+					// Quoting the docs on watch.Event.Object:
-+					//
-+					// > Object is:
-+					// >  * If Type is Added or Modified: the new state of the object
-+					// >  * If type is Deleted: the state of the object immediately before deletion.
-+					// >  * If Type is Bookmark: the object [ ... ] where only ResourceVersion field
-+					// >    is set.
-+					// >  * If Type is Error: *api.Status is recommended; other types may make sense
-+					// >    depending on context.
-+					//
-+					// So basically, we want to process the object only if ev.Type is Added,
-+					// Modified, or Deleted.
-+					if ev.Type == watch.Added || ev.Type == watch.Modified || ev.Type == watch.Deleted {
-+						pod := ev.Object.(*apiv1.Pod)
-+						updatePodRequestsFromNeonVMAnnotation(pod)
-+					}
-+
-+					// Pass along the maybe-updated event
-+					select {
-+					case <-proxyIface.StopChan():
-+						return
-+					case wrappedEvents <- ev:
-+						// continue on to next event
-+					}
-+				}
-+			}()
-+
-+			return proxyIface, nil
-+		},
-+		DisableChunking: lw.DisableChunking,
++	var usage virtualMachineUsage
++	if err := json.Unmarshal([]byte(annotation), &usage); err != nil {
++		return
 +	}
++	pod.Spec.Containers[0].Resources.Requests = apiv1.ResourceList(map[apiv1.ResourceName]resource.Quantity{
++		apiv1.ResourceCPU:    usage.CPU,
++		apiv1.ResourceMemory: usage.Memory,
++	})
 +}
 +
- // NodeLister lists nodes.
- type NodeLister interface {
- 	List() ([]*apiv1.Node, error)
+ // List returns all scheduled pods.
+ func (lister *AllPodLister) List() ([]*apiv1.Pod, error) {
+ 	var pods []*apiv1.Pod
+@@ -229,9 +252,12 @@ func (lister *AllPodLister) List() ([]*apiv1.Pod, error) {
+ 	if err != nil {
+ 		return pods, err
+ 	}
++
```

Comment:
unneeded change? (ordinarily it'd be fine, but it's worth reducing the number of lines adjacent to the diff)